app/models/concerns/wizard_form_concern.rb

```diff
@@ -0,0 +1,31 @@
+module WizardFormConcern
+  extend ActiveSupport::Concern
+
+  included { attr_accessor :form_step }
+
+  class_methods do
+    def on_wizard_step(step, exact: false, &block)
+      with_options on: :update do
+        with_options if: -> { required_for_step?(step, exact:) }, &block
+      end
+    end
+  end
```

Comment:
One thing missing here is some mechanism to force the user to define a `form_steps` method on the class. I didn't look into doing this, not sure what would be idiomatic, would it go into the `included` bit, perhaps in an initialiser?